Added absl::optional dependency to replace current optional implementation of envoy

srikailash opened the pull request:
Signed-off-by: sri kailash <sri.gandebathula@booking.com>
*Description*:
This is a work in progress pull request to replace our current optional.h with absl::optional.
refers to https://github.com/envoyproxy/envoy/issues/2651.
*Risk Level*: Low
*Testing*: Added unit test until current implementation.

htuch commented on test/common/common/abseil_optional_test.cc:
```diff
@@ -0,0 +1,18 @@
+#include "gtest/gtest.h"
+
+namespace Envoy {
+TEST(abseil_optional , basic_operations) {
```
You can skip this I think, Abseil should have its own set of tests (plus, most of this is a pretty big violation of Envoy C++ style ;) )

srikailash replied:
Sure, added this for my own checks.

ccaraman replied:
For your reference, here are the absl::optional<> tests: https://github.com/abseil/abseil-cpp/blob/master/absl/types/optional_test.cc

srikailash replied:
@ccaraman, Thanks!

htuch commented on include/envoy/common/optional.h:
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "envoy/common/exception.h"
+#include "absl/types/optional.h"
```
I rather folks just consume this directly across the tree, rather than having indirection here via our own wrapper header/build target.

srikailash replied:
ack

htuch replied:
Generally I think it's just clearer to use the actual direct includes. I think we'll eventually go with the C++17 `std::optional,` which `absl::optional` is supposed to be compatible with, so we'll just do a giant `sed` at that point.

htuch commented on source/common/router/config_impl.h:
```diff
@@ -13,6 +13,7 @@
 #include "envoy/api/v2/rds.pb.h"
 #include "envoy/api/v2/route/route.pb.h"
 #include "envoy/common/optional.h"
```
Do you want to remove the old one in the same sweep?

srikailash replied:
Well, it's still a work in progress PR. I don't mind doing that either. Let me know if you want me to do that.

htuch replied:
Sure, let's just do one giant PR, it should be fairly easy to review, as the transforms should all be mechanical. Thanks.

srikailash commented:
Let me help you with this giant review process:
Bulk changes:
      1. Optional -> abs::optional
      2. Call to .valid() function need not be done as abs::optional variables itself evaluates to true/false
      3. var.value(new_value) -> var = new_value
      4. Removed all the includes of envoy/common/optional.h
      5. Added “absl/types/optional.h”
Trivial changes (non-bulk):
	1 . Removed tests to optional.h
        2. Removed code of optional.h
Non trivial changes
        1. Added conditional return to bool functions which return .valid() call for old optional 
        2. Included exception.h directly in shared_memory_hashset.h
        3. Since cluster_impl_test expects an exception for noSdsConfig test, throwing exception code is added to retry_state_impl.cc

htuch commented:
@srikailash Thanks! This is a really nice cleanup. Can you get this to pass the failing CI checks? E.g. fix format, DCO, etc?

srikailash commented:
@htuch, Thanks for the review. 
Fixed other ci tests except format. I ran tools/check_format.py fix but i see that it still fails format.
Questions : 
1 . tools/check_format.py fix gives this error message for many BUILD files 
"sh: /bin/buildifier: No such file or directory". Anyclue what to do about this?
2 . Do we have documentation to know how to fix format tests (i found one that says to run check_format.py fix)
Thanks!

srikailash commented:
@htuch, fixed all ci tests except format. Tried to fix it using "tools/check_format.py fix" but turns out it still fails format tests. Any other way to fix format? Thanks!

htuch commented:
@srikailash did you run `check_format` under the Docker image? If not, I recommend you do, as you might have a different version of `clang-format` installed locally.

htuch commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -789,7 +789,7 @@ void ConnectionManagerImpl::startDrainSequence() {
   Router::RouteConstSharedPtr route = snapped_route_config_->route(*request_headers_, stream_id_);
   request_info_.route_entry_ = route ? route->routeEntry() : nullptr;
-  cached_route_.value(std::move(route));
+  cached_route_ = (std::move(route));
```
Can we get rid of the extra parens at sites like this?

srikailash replied:
I saw this coming ;). will update these things as well.

htuch commented on source/common/config/cds_json.cc:
```diff
@@ -140,6 +140,9 @@ void CdsJson::translateCluster(const Json::Object& json_cluster,
     cluster.set_type(envoy::api::v2::Cluster::ORIGINAL_DST);
   } else {
     ASSERT(string_type == "sds");
+    if(!eds_config.has_value()){
```
Missing spaces here...

srikailash replied:
ack

htuch commented on source/common/config/cds_json.cc:
```diff
@@ -140,6 +140,9 @@ void CdsJson::translateCluster(const Json::Object& json_cluster,
   } else {
     ASSERT(string_type == "sds");
+    if(!eds_config.has_value()){
+      throw EnvoyException("fetching invalid Optional value");
```
In the description you mention an new exception in `retry_state_impl.cc`; I don't see that, but I see this.. is this what you meant?

srikailash replied:
Yea, updated the description. Sorry about that.

htuch commented on source/common/tracing/zipkin/zipkin_core_types.h:
```diff
@@ -171,7 +171,7 @@ class Annotation : public ZipkinBase {
    * @return true if the endpoint attribute is set, or false otherwise.
    */
-  bool isSetEndpoint() const { return endpoint_.valid(); }
+  bool isSetEndpoint() const {  if(endpoint_){return true;} return false; }
```
This should just be `return endpoint_.has_value();`

srikailash replied:
Agreed.

htuch commented:
@jmarantz can you take a pass on this when the next patches arrive?

srikailash commented:
@htuch , @jmarantz fixed all changes requested in previous review. Apparently has_value() call timesout for //test/integration:http2_integration_test (https://circleci.com/gh/envoyproxy/envoy/30033?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link or please check last two commits). Thanks!

htuch commented:
@srikailash the OS X timeout is a known flake, please ignore. You're going to have to merge master to resolve the conflicts, unfortunately due to the size of this PR it's going to have to be done on an ongoing basis until we merge.

htuch commented on source/common/config/bootstrap_json.cc:
```diff
@@ -17,16 +17,17 @@ void BootstrapJson::translateClusterManagerBootstrap(
         *json_sds,
         *bootstrap.mutable_dynamic_resources()->mutable_deprecated_v1()->mutable_sds_config());
-    eds_config.value(bootstrap.dynamic_resources().deprecated_v1().sds_config());
+    eds_config = (bootstrap.dynamic_resources().deprecated_v1().sds_config());
```
Still some more surplus parens.

srikailash replied:
Will do a second pass over the changes to fix these. Thanks for the review.

htuch commented on source/common/tracing/zipkin/zipkin_core_types.h:
```diff
@@ -240,17 +246,22 @@ class BinaryAnnotation : public ZipkinBase {
    */
-  bool isSetEndpoint() const { return endpoint_.valid(); }
+  bool isSetEndpoint() const {
+    if (endpoint_) {
```
I assume you'll replace with `has_value` given the comment re: OS X flake? Thanks.

srikailash replied:
Sure, will just have to revert the last commit.

htuch replied:
This one still needs fixing.

srikailash commented:
@htuch, i probably need some help to resolve conflicts for some files. Do you think copying file from upstream and then re-applying replacement changes would be a good idea?(I am new to big code bases). Thanks!

jmarantz commented:
I have some general suggestions about how to make large scale refactors easier to push through.

1. Automate the changes if you can, via clang or sed or whatever, and review -- or at least describe it in the PR description, which might be easier to review than the results of the automation.
2. Try to make the changes incremental with every step obviously &/or proveably correct
3. Break down the changes into a few reviewable chunks at a time, which makes it easier to push through review/merge without conflicts.

For this case, I'm not sure about suggestion 1 above.  Suggestion 2 might be a lot of work if you don't do suggestion 1, but a thought here is to:
 - start by adding absl::optional's interface into Envoy's Optional class, and converting call-sites over in reviewable swaths
 - When done, remove the class and replace it with a 'using' to alias it to absl::optional (which should be a tiny change)
 -  Then do one another sweep through the codebase and change all the references, so you can finally remove the alias.

Each of these changes should be much easier to review and merge without conflicts.

But at a minimum, it doesn't seem like it's necessary for you to remove the existing Optional class all at once, so you could just revert the conflicting files and do more passes later to clean them up.

htuch commented:
@jmarantz I asked for the all-in-one PR here, as it's fairly mechanical IMHO.

htuch commented:
Ideally you have a script to automate the `sed`(or the like as suggested by @jmarantz), so you should be able to either resolve conflicts or start in a new client and do this. If not, you might want to do what you suggest.

srikailash commented:
@htuch, @jmarantz thanks for the suggestions.
@htuch apparently applying replacements on upstream code for conflicting files and falling back to old Optional in some files is not as smooth as i assumed(many not viable conversion errors are popping up). Mind if i close this PR and submit a new one with steps suggested by @jmarantz later?
Thanks!

srikailash commented on test/common/common/BUILD:
```diff
@@ -10,13 +10,6 @@ load(
 
 envoy_package()
 
-envoy_cc_fuzz_test(
```
Hello @htuch, fixed merge conflicts and made sure there are no extra braces. 
Adding this test gives me this error "  //test/common/common:base64_fuzz_test_lib: invalid label 'base64_corpus/.!31337!favicon.ico' in element 0 of attribute 'data' in 'cc_library' rule: invalid target name 'base64_corpus/.!31337!favicon.ico': target names may not contain '!'
 ". Any idea how to fix this? Thanks in advance!

htuch replied:
Look at your diff below, it looks like you're trying to check-in some weird stuff like `test/common/common/base64_corpus/.!31337!favicon.ico`. Can you `git rm` these from your repo and try again? We do need to preserve this target :)

htuch commented on source/common/config/cds_json.cc:
```diff
@@ -140,6 +140,9 @@ void CdsJson::translateCluster(const Json::Object& json_cluster,
     cluster.set_type(envoy::api::v2::Cluster::ORIGINAL_DST);
   } else {
     ASSERT(string_type == "sds");
+    if (!eds_config.has_value()) {
```
Technically this can be `if (!eds_config)` I think.

htuch commented on source/common/router/config_impl.cc:
```diff
@@ -151,17 +151,18 @@ HashPolicyImpl::HashPolicyImpl(
-      const uint64_t old_value = hash.valid() ? ((hash.value() << 1) | (hash.value() >> 63)) : 0;
-      hash.value(old_value ^ new_hash.value());
+      const uint64_t old_value = hash ? ((hash.value() << 1) | (hash.value() >> 63)) : 0;
+      hash = (old_value ^ new_hash.value());
```
Nit: can remove parens here.

htuch commented on source/common/router/config_utility.h:
```diff
@@ -108,16 +109,17 @@ class ConfigUtility {
-   *         or the HTTP status code from the route's redirect if specified,
-   *         or an empty Option otherwise.
+   * @return absl::optional<Http::Code> the HTTP status from the route's direct_response if
+   * specified, or the HTTP status code from the route's redirect if specified, or an empty Option
```
Nit: vestigial reference to Option.

srikailash commented on source/common/tracing/zipkin/zipkin_core_types.h:
```diff
@@ -248,16 +248,6 @@ class BinaryAnnotation : public ZipkinBase {
-  /**
-   * @return true of the endpoint attribute has been set, or false otherwise.
-   */
-  bool isSetEndpoint() const {
```
I see that this function is defined twice. Removed one instance.

srikailash commented:
@htuch , Thanks for the review. Fixed things mentioned by you.

srikailash commented on source/common/config/cds_json.cc:
```diff
@@ -140,6 +140,9 @@ void CdsJson::translateCluster(const Json::Object& json_cluster,
     cluster.set_type(envoy::api::v2::Cluster::ORIGINAL_DST);
   } else {
     ASSERT(string_type == "sds");
+    if (!eds_config) {
```
@mattklein123 , copying from what i mentioned earlier about this change.
"Since cluster_impl_test expects an exception for noSdsConfig test, throwing exception code is added to cds_json.cc." (our previous optional.h throws exception when trying to retrieve value from empty optional.)
Thanks for the review!

mattklein123 replied:
OK I guess that was a bad error message before. Since you are already in here, can you potentially change the error message to be more useful and change the test to use EXPECT_THROW_WITH_MESSAGE to check it? Thank you!

srikailash replied:
@mattklein123, fixed. please take a look. Thanks!

srikailash commented:
Thanks for merging :) will pick up something else from issues now.
